design/open/delta-diff.md

```diff
@@ -0,0 +1,168 @@
+2. `LAKEFS_PLUGIN_DIFF_{TYPE OF DIFF}`
+   The DiffService will use the name of binary provided by this env var to perform the diff. For instance, `LAKEFS_PLUGIN_DIFF_DELTA=deltaDiffBinary`
+   The DiffService will use the [plugins' location](https://github.com/hashicorp/terraform/blob/main/plugins.go) and the diff binary to load the plugin and request a diff.
+   The **type** of diff will be sent as part of the request to lakeFS as specified [here](#API).
+   The communication between the DiffService and the plugins, as explained above, will be through the `go-plugin` package (`gRPC`).
+
+#### Delta Diff Plugin
+
+Implemented using [delta-rs](https://github.com/delta-io/delta-rs) (Rust), this plugin will perform the diff operation using table paths provided by the DiffService through a `gRPC` call.  
+To query the Delta Table from lakeFS, the plugin will generate an S3 client (this is a constraint imposed by the `delta-rs` package) and send a request to lakeFS's S3 gateway.  
+The diff algorithm:
+1. Run the Delta [HISTORY command](https://docs.delta.io/latest/delta-utility.html#history-schema) on both table paths.
+2. Traverse through the [returned "commitInfo" entry vector ](https://github.com/delta-io/delta-rs/blob/main/rust/src/delta.rs#L888)
+starting from the **last** version of each history vector: 
+    1. While the returned entry versions **aren't** equal:
+        1. If the bigger version is the _"left"'s_ version, add the entry to the returned history list.
+        2. Traverse one version back of the bigger version entry's history.
+    2. While the versions > 0: 
+        1. Create a hash for the entry based on fields: `timestamp`, `operation`, `operationParameters` , and `operationMetrics` values.
+        2. Compare the hashes of the versions.
+        3. If they **aren't equal**, add the "left"'s entry to the returned history list, else break and **return the history vector**.
+        4. Traverse one version back in both vectors.
+3. Return an empty history vector.
+
+### Authentication
+
+The `delta-rs` package generates an S3 client which will be used to communicate back to lakeFS (through the S3 gateway).  
+In order for the S3 client to communicate with lakeFS (or S3 in general) it needs to pass an AWS Access Key Id and Secret access key.  
+Since applicative credentials are not obligatory, the users that sent the request for a diff might not have such, and even if they have,
+they cannot send them through the GUI (which is the UI we chose to implement this feature for the MVP).
+To overcome this scenario, we'll use special diff credentials as follows:
+1. User makes a Delta Diff request.
+2. The DiffService checks if the user has "diff credentials" in the DB:
+    1. If there are such credentials, it will use them.
+    2. If there aren't such, it will generate the credentials and save them: `{AKIA: DIFF-<>, SAK: <>}`. The `DIFF` prefix will be used to identify "diff credentials".
```

Comment:
1. We should get someone from the auth team to go over this!
1. I would like us to declare this solution is explicitly temporary.  There are at least 2 auth requirements in here:
   * Support a note on credentials
   * Allow temporary credentials.
   * (Possibly...) allow attaching a policy with reduced permissions to such credentials.

Reply:
1. This solution involved @guy-har. Adding as a reviewer...
2. Absolutely.

Reply:
Generate for missing is good short term solution - it can also be an issue when two services will try to generate while currently we don't lock this option means we will have two sets of credentials.

Reply:
Ok,
I'll mark the temporary solutions (for the MVP) all around the document.